### DISCUSSION TITLE
How to set per-user ACLs?

### DISCUSSION THREAD

**Discussion by db0** [django-wiki/django-wiki#1308] [Q&A] [answered]
In the settings I see values to set permissions for all users, but it's not clear to me how I can set permissions per user. For example I want be default everyone to have read-access, but I would provide write access on demand.

**Comment by benjaoming** [comment]
This is handled in the wiki itself.

Try to have a look at the demo, login with admin:admin

https://demo.django-wiki.org/

**Comment by db0** [comment]
I get that there's permissions in the wiki, but I don't understand how to use them. I want to prevent new users from being able to modify the wiki, to avoid spammers. I have set `WIKI_CAN_WRITE = False` in `settings.py` but new users apparently can still create and edit.

I already have a wiki running now btw.

I discovered I can set some ACLs via `/admin/auth/user/` so I could use this after a user registers to assign them to groups, but I can't figure out how to prevent them from having write access by default.

**Comment by db0** [comment]
Ah I see, so I need to change permissions to the top article itself, to prevent `others` from writing

**Reply by benjaoming** [reply]
Exactly - articles have permissions like files in a file system. There is no global switch for users, but you can implement your own callable that accepts arguments `(article, user)`

https://github.com/django-wiki/django-wiki/blob/918b354aabf23ae7c1612e9ac746d7c7905c137b/src/wiki/core/permissions.py#L43-L60

**Comment by db0** [comment]
Interesting. I am planning an attempt to make the django-wiki a partner to my lemmy fediverse site (https://lemmy.dbzer0.com), so I want an easy way only for my lemmy users to be able to write. 

I could easily make a python method that checks if said username exists in my lemmy instance. Could you guide me me how I could implement this callable? I assume I would put it as a python file somewhere but how do I then hook it up to the wiki?

**Reply by db0** [reply]
I've started writing the callable. Would the username for the user be `user.username` or something else?

would the py file go to the same place as `settings.py`, or somewhere else?

**Reply by db0** [reply]
I really can't figure out where to put my `callables.py`.

**Reply by db0** [reply, marked as answer]
Ah, I figured it out. I need to import it in settings.py

`from .callables import is_write_user,is_admin_user`

**Reply by benjaoming** [reply]
Exactly! And this is the perfect usecase for writing your own callable :+1:

**Reply by db0** [reply]
Yes it's brilliant! You're about to get a lot more popular as I expect a lot of fediverse instances might copy my approach once I publish it :D

**Reply by benjaoming** [reply]
It would be super-nice to broaden the usage of Django wiki, and it's really nice to see the energy around Lemmy and federation in general. Wishing you a lot of success :pray: :rainbow:

**Reply by db0** [reply]
Deployed here: https://wiki.dbzer0.com/

Ignore the cert issue for now
